ownCloud File Provider/FileProviderContentEnumerator.m

```diff
@@ -0,0 +1,687 @@
+			dispatch_async(dispatchQueue, ^{
+				job(completionHandler);
+			});
+		};
+	});
+
+	return (queue);
+}
+
+#pragma mark - Initialization
+- (instancetype)initWithVFSCore:(OCVFSCore *)vfsCore containerItemIdentifier:(OCVFSItemID)containerItemIdentifier;
+{
+	if ((self = [super init]) != nil)
+	{
+		_vfsCore = vfsCore;
+		_containerItemIdentifier = containerItemIdentifier;
+
+		_enumerationObservers = [NSMutableArray new];
+		_changeObservers = [NSMutableArray new];
+
+		[[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(_displaySettingsChanged:) name:DisplaySettingsChanged object:nil];
+	}
+
+	return (self);
+}
+
+#pragma mark - Display settings change tracking
+- (void)_displaySettingsChanged:(NSNotification *)notification
+{
+	OCLogDebug(@"Received display settings update notification (enumerator for %@)", _containerItemIdentifier);
+
+	if (_content.query != nil)
+	{
+		[DisplaySettings.sharedDisplaySettings updateQueryWithDisplaySettings:_content.query];
+	}
+
+	if (_containerItemIdentifier != nil)
+	{
+		[_content.core.vault signalEnumeratorForContainerItemIdentifier:_containerItemIdentifier];
+	}
+}
+
+#pragma mark - FileProvider API
+- (void)enumerateItemsForObserver:(id<NSFileProviderEnumerationObserver>)observer startingAtPage:(NSFileProviderPage)page
+{
+	OCLogDebug(@"##### Enumerate ITEMS for observer: %@ fromPage: %@", observer, page);
+
+	__weak FileProviderContentEnumerator *weakSelf = self;
+
+	[self requestContentWithErrorHandler:^(NSError *error) {
+		dispatch_async(dispatch_get_main_queue(), ^{
+			[observer finishEnumeratingWithError:error];
+		});
+	} contentConsumer:^(OCVFSContent *content) {
+		dispatch_async(dispatch_get_main_queue(), ^{
+			[weakSelf provideItemsToEnumerationObserver:observer fromContent:content];
+		});
+	} observerQueuer:^BOOL(dispatch_block_t completionHandler) {
+		FileProviderContentEnumerator *strongSelf = weakSelf;
+
+		if (strongSelf == nil) {
+			return(NO);
+		}
+
+		FileProviderEnumeratorObserver *enumerationObserver = [FileProviderEnumeratorObserver new];
+
+		enumerationObserver.enumerationObserver = observer;
+		enumerationObserver.enumerationStartPage = page;
+		enumerationObserver.enumerationCompletionHandler = completionHandler;
+
+		[strongSelf->_enumerationObservers addObject:enumerationObserver];
+
+		return (YES);
+	}];
+
+	/* TODO:
+	- inspect the page to determine whether this is an initial or a follow-up request
+
+	If this is an enumerator for a directory, the root container or all directories:
+	- perform a server request to fetch directory contents
+	If this is an enumerator for the active set:
+	- perform a server request to update your local database
+	- fetch the active set from your local database
+
+	- inform the observer about the items returned by the server (possibly multiple times)
+	- inform the observer that you are finished with this page
+	*/
+}
```

Comment:
Those are the comments from Apple's template. Since FileProvider documentation spreads out in WWDC videos, header files, developer documentation and those comments, I'd like to keep them around for reference.